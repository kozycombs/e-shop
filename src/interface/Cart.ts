export interface CartProduct {
  productId: number;
  quantity: number;
}

export interface Cart {
  id: number;
  userId: number;
  date: string;
  products: CartProduct[];
}

export interface CartResponse {
  id: number;
}
