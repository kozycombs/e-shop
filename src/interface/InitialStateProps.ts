export interface InitialStateProps<T> {
  data: T;
  error: string;
}
