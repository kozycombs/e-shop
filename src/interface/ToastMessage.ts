export interface ToastMessage {
  message: string;
  severity: "success" | "error";
}
