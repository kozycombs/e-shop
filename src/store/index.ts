import { configureStore } from "@reduxjs/toolkit";
import productReducer from "store/productSlice";
import cartReducer from "store/cartSlice";

export const store = configureStore({
  reducer: {
    products: productReducer,
    cart: cartReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
