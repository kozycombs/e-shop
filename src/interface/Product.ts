import { Rating } from "interface/Rating";

export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
  rating: Rating;
}
