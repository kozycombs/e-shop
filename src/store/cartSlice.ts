import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { RootState } from ".";
import { API_BASE_URL } from "../constants";
import { InitialStateProps } from "../interface/InitialStateProps";
import { Cart, CartProduct } from "../interface/Cart";

export const initialState: InitialStateProps<Cart | null> = {
  loading: false,
  data: null,
  error: "",
};

export const cartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    updateCart: (state, action: { payload: Cart }) => {
      state.data = action.payload;
    },
    resetCart: () => {
      return { ...initialState };
    },
  },
  extraReducers(builder) {
    builder
      .addCase(addToCart.fulfilled, (state, action) => {
        state.data = action.payload;
      })
      .addCase(updateCartProduct.fulfilled, (state, action) => {
        state.data = action.payload;
      });
  },
});

export const { updateCart, resetCart } = cartSlice.actions;

export default cartSlice.reducer;

export const cartSelector = (state: RootState) => state.cart;

export const addToCart = createAsyncThunk(
  "cart/addToCart",
  async ({
    userId,
    date,
    products,
  }: {
    userId: number;
    date: string;
    products: CartProduct[];
  }): Promise<Cart> => {
    const response = await fetch(`${API_BASE_URL}/carts`, {
      method: "POST",
      body: JSON.stringify({
        userId,
        date,
        products,
      }),
    });
    const result = await response.json();
    return {
      id: Number(result.id),
      userId,
      date,
      products,
    };
  }
);

export const updateCartProduct = createAsyncThunk(
  "cart/updateCartProduct",
  async ({
    cartId,
    userId,
    date,
    products,
  }: {
    cartId: number;
    userId: number;
    date: string;
    products: CartProduct[];
  }): Promise<Cart> => {
    const response = await fetch(`${API_BASE_URL}/carts/${cartId}`, {
      method: "PUT",
      body: JSON.stringify({
        userId,
        date,
        products,
      }),
    });
    const result = await response.json();
    return {
      id: Number(result.id),
      userId,
      date,
      products,
    };
  }
);
