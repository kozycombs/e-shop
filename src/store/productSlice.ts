import { createSlice } from "@reduxjs/toolkit";
import { RootState } from ".";
import { Product } from "../interface/Product";

export const initialState: Product[] = [];

export const productSlice = createSlice({
  name: "products",
  initialState,
  reducers: {
    updateProducts: (state, action) => {
      state = action.payload;
    },
  },
});

export const { updateProducts } = productSlice.actions;

export default productSlice.reducer;

export const productsSelector = (state: RootState) => state.products;
