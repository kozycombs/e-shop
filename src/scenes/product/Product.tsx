import { FC, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import {
  Alert,
  Box,
  Button,
  Divider,
  FormControl,
  Grid,
  MenuItem,
  Rating,
  Select,
  SelectChangeEvent,
  Snackbar,
  Typography,
} from "@mui/material";
import { fetchProduct, productsSelector } from "store/productSlice";
import { Product as ProductInterface } from "interface/Product";
import Spinner from "components/spinner/Spinner";
import { addToCart, cartSelector, updateCartProduct } from "store/cartSlice";
import { AppDispatch } from "store/index";
import { ToastMessage } from "interface/ToastMessage";
import { PRODUCT_QUATITY_COUNT } from "constants/index";
import ErrorBanner from "components/errorBanner/ErrorBanner";

const Product: FC = () => {
  const { id } = useParams();
  const dispatch = useDispatch<AppDispatch>();
  const { data: products } = useSelector(productsSelector);
  const { data: cart } = useSelector(cartSelector);
  const [loading, setLoading] = useState(false);
  const [product, setProduct] = useState<ProductInterface>();
  const [quantity, setQuantity] = useState(1);
  const [toastMessage, setToastMessage] = useState<ToastMessage>({
    message: "",
    severity: "success",
  });
  const [error, setError] = useState("");
  const userId = 1; // This is a dummy user id, since there is no logged in functionality.
  const vertical = "top";
  const horizontal = "right";

  useEffect(() => {
    const getProduct = async () => {
      try {
        const result = await dispatch(fetchProduct(Number(id))).unwrap();
        setProduct(result);
        setLoading(false);
      } catch (error) {
        setError("Failed to load our product. Please try again later");
        setLoading(false);
      }
    };

    // If product list is empty get product detail from API
    // otherwiser get product detail from cached product list.

    if (products.length <= 0) {
      setLoading(true);
      getProduct();
    } else {
      setLoading(true);
      const product =
        products.find((currentProduct) => currentProduct.id === Number(id)) ||
        null;
      if (product) {
        setProduct(product);
      }
      setLoading(false);
    }
  }, [id, products, dispatch]);

  const resetToastMessage = () => {
    setToastMessage({ message: "", severity: "success" });
  };

  const handleQuantityChange = (event: SelectChangeEvent) => {
    setQuantity(Number(event.target.value));
  };

  const handleAddToCart = async () => {
    if (!product) {
      return;
    }

    try {
      setLoading(true);
      const date = new Date().toISOString();

      // If cart is null add to cart else update with cart id.
      if (!cart) {
        const result = await dispatch(
          addToCart({
            userId,
            date,
            products: [{ productId: product?.id, quantity }],
          })
        ).unwrap();
        if (result.id) {
          setToastMessage({
            message: "Product added to cart successfully",
            severity: "success",
          });
        } else {
          setToastMessage({
            message: "Failed to add product to cart",
            severity: "error",
          });
        }
      } else {
        const productFoundIndex = cart.products.findIndex(
          (currentProduct) => currentProduct.productId === product.id
        );

        if (productFoundIndex > -1) {
          const updatedProduct = {
            ...cart.products[productFoundIndex],
            quantity: cart.products[productFoundIndex].quantity + quantity,
          };
          const otherProducts = cart.products.filter(
            (currentProduct) => currentProduct.productId !== product.id
          );
          const products = [...otherProducts, updatedProduct];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        } else {
          const products = [
            ...cart.products,
            { productId: product?.id, quantity },
          ];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        }
      }
      setLoading(false);
    } catch (error) {
      setError("Failed to add product to cart. Please try again later");
      setLoading(false);
    }
  };

  return (
    <Box sx={{ pt: 4 }}>
      {loading && (
        <Box className="flex justify-center">
          <Spinner />
        </Box>
      )}

      {error.length > 0 && <ErrorBanner message={error} />}

      <Snackbar
        anchorOrigin={{ vertical, horizontal }}
        open={toastMessage?.message.length > 0}
        autoHideDuration={2000}
        onClose={resetToastMessage}
        key={vertical + horizontal}
      >
        <Alert
          onClose={resetToastMessage}
          severity={toastMessage?.severity}
          variant="filled"
          sx={{ width: "100%" }}
        >
          {toastMessage?.message}
        </Alert>
      </Snackbar>

      <Grid container spacing={{ xs: 4, md: 3 }} columns={{ xs: 1, sm: 4 }}>
        <Grid item xs={1} sm={2} sx={{ marginBottom: 10 }}>
          <Box
            component="img"
            sx={{
              height: "auto",
              width: "100%",
              maxHeight: { xs: 400 },
              objectFit: "contain",
            }}
            alt={product?.title}
            src={product?.image}
          />
        </Grid>
        <Grid item xs={1} sm={2}>
          <Box></Box>
          <Typography
            variant="h1"
            sx={{ mb: 1, fontWeight: "400", fontSize: 24 }}
          >
            {product?.title}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "end", mb: 2 }}>
            <Typography
              component="span"
              sx={{
                mr: 1,
                fontSize: 16,
              }}
            >
              {product?.rating.rate}
            </Typography>
            <Rating
              name="read-only"
              value={Number(product?.rating.rate)}
              readOnly
            />
            <Typography
              component="span"
              sx={{
                ml: 1,
                mt: 1,
                fontSize: 16,
                textDecoration: "underline",
              }}
            >
              {product?.rating.count} reviews
            </Typography>
          </Box>
          <Typography
            component="p"
            sx={{ fontSize: 24, fontWeight: 400, mb: 4 }}
          >
            &pound;{Number(product?.price).toFixed(2)}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "center", mb: 4 }}>
            <FormControl sx={{ m: 1, minWidth: 120 }} size="small">
              <Select
                id="quantity"
                data-test="quantity"
                value={`${quantity}`}
                onChange={handleQuantityChange}
              >
                {Array(PRODUCT_QUATITY_COUNT)
                  .fill("_")
                  .map((_, index) => (
                    <MenuItem key={index} value={index + 1}>
                      {index + 1}
                    </MenuItem>
                  ))}
              </Select>
            </FormControl>
            <Button
              variant="contained"
              size="medium"
              disableElevation
              onClick={handleAddToCart}
              sx={{ textTransform: "none" }}
            >
              Add to Cart
            </Button>
          </Box>
          <Divider />
          <Typography
            component="p"
            sx={{
              mb: 2,
              mt: 4,
              fontSize: 16,
            }}
          >
            {product?.description}
          </Typography>
        </Grid>
      </Grid>
    </Box>
  );
};

export default Product;
