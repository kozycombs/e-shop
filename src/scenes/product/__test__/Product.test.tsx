import { Provider } from "react-redux";
import { render, screen, configure } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { updateProducts } from "store/productSlice";
import { store } from "store/index";
import { Product } from "interface/Product";
import ProductDetail from "scenes/product/Product";

describe("<ProductDetail />", () => {
  beforeAll(() => {
    configure({
      testIdAttribute: "data-test",
    });
    const product: Product = {
      id: 16,
      title:
        "Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket",
      price: 29.95,
      description:
        "100% POLYURETHANE(shell) 100% POLYESTER(lining) 75% POLYESTER 25% COTTON (SWEATER), Faux leather material for style and comfort / 2 pockets of front, 2-For-One Hooded denim style faux leather jacket, Button detail on waist / Detail stitching at sides, HAND WASH ONLY / DO NOT BLEACH / LINE DRY / DO NOT IRON",
      category: "women's clothing",
      image: "https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg",
      rating: {
        rate: 2.9,
        count: 340,
      },
    };

    store.dispatch(updateProducts([product]));
  });

  it("should render the component", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ProductDetail />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have a quantity input field", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ProductDetail />
        </BrowserRouter>
      </Provider>
    );

    const quantityInput = screen.getByTestId("quantity");
    expect(quantityInput).toBeInTheDocument();
  });

  it("should have a Add to Cart button", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ProductDetail />
        </BrowserRouter>
      </Provider>
    );

    const addToCartButton = screen.getByRole("button");
    expect(addToCartButton.textContent).toBe("Add to Cart");
  });
});
