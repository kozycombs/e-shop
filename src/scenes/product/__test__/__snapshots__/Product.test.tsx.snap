// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetail /> should render the component 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-xuhmcu"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-spacing-md-3 css-1rs5ils-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-2 css-1e4fi12-MuiGrid-root"
      >
        <img
          class="MuiBox-root css-1y2a96z"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-2 css-wdtwov-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-0"
        />
        <h1
          class="MuiTypography-root MuiTypography-h1 css-28f9h8-MuiTypography-root"
        />
        <div
          class="MuiBox-root css-glidez"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 css-1l14lsq-MuiTypography-root"
          />
          <span
            aria-label="NaN Stars"
            class="MuiRating-root MuiRating-sizeMedium Mui-readOnly MuiRating-readOnly css-15befwp-MuiRating-root"
            role="img"
          >
            <span>
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                  data-testid="StarBorderIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                  data-testid="StarBorderIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                  data-testid="StarBorderIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                  data-testid="StarBorderIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                  data-testid="StarBorderIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                  />
                </svg>
              </span>
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 css-it9kvy-MuiTypography-root"
          >
             reviews
          </span>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 css-dh0nom-MuiTypography-root"
        >
          £NaN
        </p>
        <div
          class="MuiBox-root css-1ol10sz"
        >
          <div
            class="MuiFormControl-root css-1869usk-MuiFormControl-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              data-test="quantity"
            >
              <div
                aria-controls=":r6:"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="quantity"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="quantity"
                role="combobox"
                tabindex="0"
              >
                1
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="1"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation css-1hx2tqu-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Add to Cart
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1py8l59-MuiTypography-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
